airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/DestinationTaskScopeProvider.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.task
+
+import io.airbyte.cdk.command.DestinationConfiguration
+import io.github.oshai.kotlinlogging.KotlinLogging
+import io.micronaut.context.annotation.Secondary
+import jakarta.inject.Singleton
+import java.util.concurrent.Executors
+import kotlin.system.measureTimeMillis
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.supervisorScope
+import kotlinx.coroutines.withTimeoutOrNull
+
+/**
+ * The scope in which a task should run
+ * - InternalTask:
+ * ```
+ *       - internal to the task launcher
+ *       - should not be blockable by implementor errors
+ *       - killable w/o side effects
+ * ```
+ * - ImplementorTask: implemented by the destination
+ * ```
+ *       - calls implementor interface
+ *       - should not block internal tasks (esp reading from stdin)
+ *       - should complete if possible even when failing the sync
+ * ```
+ */
+sealed interface ScopedTask : Task
+
+interface InternalTask : ScopedTask
+
+interface ImplementorTask : ScopedTask
+
+@Singleton
+@Secondary
+class DestinationTaskScopeProvider(config: DestinationConfiguration) :
+    TaskScopeProvider<ScopedTask> {
+    private val log = KotlinLogging.logger {}
+
+    private val timeoutMs = config.gracefulCancellationTimeoutMs
+
+    data class ControlScope(
+        val job: Job,
+        val dispatcher: CoroutineDispatcher,
+        val scope: CoroutineScope = CoroutineScope(dispatcher + job)
```

Comment:
for my knowledge: what does `dispatcher + job` do?

Reply:
Adding a job let's you cancel/join all jobs in the scope by canceling/joining the job.